flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/table/KinesisDynamicTableFactoryTest.java

```diff
@@ -0,0 +1,358 @@
+
+	// --------------------------------------------------------------------------------------------
+	// Positive tests
+	// --------------------------------------------------------------------------------------------
+
+	@Test
+	public void testGoodTableSource() {
+		TableSchema sourceSchema = defaultSourceSchema().build();
+		Map<String, String> sourceOptions = defaultSourceOptions().build();
+
+		// Construct actual DynamicTableSource using FactoryUtil
+		KinesisDynamicSource actualSource = (KinesisDynamicSource)
+			FactoryUtil.createTableSource(
+				null,
+				ObjectIdentifier.of("default", "default", SOURCE_TABLE),
+				createSourceTable(sourceSchema, sourceOptions, Collections.emptyList()),
+				new Configuration(),
+				Thread.currentThread().getContextClassLoader(),
+				false);
+
+		// Construct expected DynamicTableSink using factory under test
+		KinesisDynamicSource expectedSource = new KinesisDynamicSource(
+			sourceSchema.toPhysicalRowDataType(),
+			SOURCE_STREAM,
+			KINESIS_PROPERTIES,
+			new TestFormatFactory.DecodingFormatMock(",", true));
+
+		// verify that the constructed DynamicTableSink is as expected
+		assertEquals(expectedSource, actualSource);
+
+		// verify that the copy of the constructed DynamicTableSink is as expected
+		assertEquals(expectedSource, actualSource.copy());
+
+		// verify produced sink
+		ScanTableSource.ScanRuntimeProvider functionProvider =
+			actualSource.getScanRuntimeProvider(ScanRuntimeProviderContext.INSTANCE);
+		SourceFunction<RowData> sourceFunction =
+			as(functionProvider, SourceFunctionProvider.class).createSourceFunction();
+		assertThat(sourceFunction, instanceOf(FlinkKinesisConsumer.class));
+	}
+
+	@Test
+	public void testGoodTableSourceWithMetadataFields() {
+		TableSchema sourceSchema = defaultSourceSchema().build();
+		Map<String, String> sourceOptions = defaultSourceOptions().build();
+
+		// Construct actual DynamicTableSource using FactoryUtil
+		KinesisDynamicSource actualSource = (KinesisDynamicSource)
+			FactoryUtil.createTableSource(
+				null,
+				ObjectIdentifier.of("default", "default", SOURCE_TABLE),
+				createSourceTable(sourceSchema, sourceOptions, Collections.emptyList()),
+				new Configuration(),
+				Thread.currentThread().getContextClassLoader(),
+				false);
+
+		// Construct expected DynamicTableSink using factory under test
+		KinesisDynamicSource expectedSource = new KinesisDynamicSource(
+			sourceSchema.toPhysicalRowDataType(),
+			SOURCE_STREAM,
+			KINESIS_PROPERTIES,
+			new TestFormatFactory.DecodingFormatMock(",", true));
+
+		List<String> metadataKeys = Arrays.asList("ShardId", "ApproximateArrivalTimestamp");
+		DataType producedDataType = getProducedType(sourceSchema, Metadata.values());
+
+		expectedSource.applyReadableMetadata(metadataKeys, producedDataType);
+		actualSource.applyReadableMetadata(metadataKeys, producedDataType);
+
+		// verify that the constructed DynamicTableSink is as expected
```

Comment:
copy past error `DynamicTableSink` here and below

Reply:
Done.